shell/shared/grain.js

```diff
@@ -1148,52 +1312,68 @@ Router.map(function () {
       ];
     },
 
-    data: function() {
-      if (!this.ready || Meteor.loggingIn()) {
-        return;
-      }
-      if (Meteor.userId() && !Session.get("visit-token-" + this.params.token)) {
-        var tokenInfo = TokenInfo.findOne(this.params.token);
-        if (!tokenInfo || !tokenInfo.apiToken) {
-          this.state.set("error", "invalid authorization token");
-        } else {
-          this.state.set("error", undefined);
-          var apiToken = tokenInfo.apiToken;
-          if (!Grains.findOne({_id: apiToken.grainId, userId: Meteor.userId()}) &&
-              !ApiTokens.findOne({userId: apiToken.userId, "owner.user.userId": Meteor.userId()})) {
+    onRun: function () {
+      var token = this.params.token;
+      var path = this.params.path;
+      var query = this.params.query;
+      var hash = this.params.hash;
+
+      Tracker.autorun(function() {
```

Comment:
I need TokenInfo to arrive, and I also need Grains to arrive, and I need ApiTokens to arrive (because they may prevent me from needing to prompt the user with an interstitial), and I need to make sure that GrainView is marked active once it's created, and I need to open a session, but only if either the user has answered the interstitial or the Grain/ApiToken tables prove it's not needed.